zio/bzngio/types.go

```diff
@@ -0,0 +1,97 @@
+// Package bzngio provides an API for reading and writing zng values and
+// directives in binary zng format.  The Reader and Writer types implement the
+// the zbuf.Reader and zbuf.Writer interfaces.  Since these methods
+// read and write only zbuf.Records, but the bzng format includes additional
+// functionality, other methods are available to read/write zng comments
+// and include virtual channel numbers in the stream.  Virtual channels
+// provide a way to indicate which output of a flowgraph a result came from
+// when a flowgraph computes multiple output channels.  The bzng values in
+// this zng value are "machine format" as prescirbed by the ZNG spec.
+// The vanilla zbuf.Reader and zbuf.Writer implementations ignore application-specific
+// payloads (e.g., channel encodings).
+package bzngio
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/mccanne/zq/zng/resolver"
+)
+
+// TypeFile manages the mapping between small-integer descriptor identifiers
+// and zng descriptor objects, which hold the binding between an identifier
+// and a zeek.TypeRecord.
+type TypeFile struct {
+	mu sync.Mutex
```

Comment:
resolver.Context already contains a mutex. Just use that?

Reply:
With two locks, other threads can be updating the type context while the type file is being written.